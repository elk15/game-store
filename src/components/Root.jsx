import Icon from '@mdi/react';
import { mdiCartOutline } from '@mdi/js';
import { mdiMagnify } from '@mdi/js';
import { mdiClose } from '@mdi/js';
import { mdiMenu } from '@mdi/js';
import { NavLink, Outlet } from 'react-router-dom';
import { useState } from 'react';
import PropTypes from 'prop-types';
import { useEffect } from 'react';

const Root = ({cart, allGames, addItemToCart}) => {
  const [isCartOpen, setIsCartOpen] = useState(false);
  const [isSearchBarOpen, setIsSearchBarOpen] = useState(false);
  const [isMobileMenuOpen, setIsMobileMenuOpen] = useState(false);
  const [searchInput, setSearchInput] = useState('');
  const [searchResults, setSearchResults] = useState([]);

  const toggleCart = () => {
    setIsCartOpen(isCartOpen ? false : true);
  }

  const closeCart = () => {
    setIsCartOpen(false);
  }

  const openSearchBar = () => {
    setIsSearchBarOpen(true);
  }

  const closeSearchBar = () => {
    setIsSearchBarOpen(false);
    setSearchInput('');
  }

  const openMobileMenu = () => {
    setIsMobileMenuOpen(true);
  }

  const closeMobileMenu = () => {
    setIsMobileMenuOpen(false);
  }

  const handleSearchInputChange = (e) => {
    setSearchInput(e.target.value);
  }

  

  useEffect(() => {
    if (searchInput.length > 1) {
      const searchForGames = (query) => {
        return allGames.filter((game) => game.name.toLowerCase().includes(query.toLowerCase()));
      }
      setSearchResults(searchForGames(searchInput));
    }

    if (searchInput.length < 2) {
      setSearchResults([]);
    }

  }, [searchInput, allGames])

  return (
    <>
      <header className='z-[100] opacity-95 flex justify-center bg-neutral-900  py-4 px-3 text-neutral-300 sticky top-0 text-sm '>
        <div className="max-w-screen-lg flex flex-1 gap-3 items-center justify-between">
          {isSearchBarOpen ?
          <>
          <div  data-testid = "overlay" className='z-[101] fixed top-0 left-0 w-full h-full' onClick={closeSearchBar}></div>
          <div className='z-[102] flex flex-1 gap-2 items-center animate-open relative'>
            <Icon path={mdiMagnify} size={1} />
            <input type="search" autoComplete='off' name="gameSearch" value={searchInput} onChange={handleSearchInputChange}
            className='w-full bg-transparent border-b-neutral-700 border-b-[1px]
            focus:outline-none'></input>
          </div>
          {searchInput.length > 1 &&
          <div className='z-[102] lg:absolute lg:top-[60px] top-[55px] text-black bg-white p-3 animate-openfast
          flex flex-col lg:max-w-[500px] rounded fixed left-0 lg:left-auto w-full h-full lg:h-auto'>
            <h2 className='font-semibold text-xl mb-2'>
              {searchResults.length} Games found
            </h2>
            <ul>
              {searchResults.map((game) => {
                return  <li key={game.id} data-id={game.id} onClick={addItemToCart}
                className='flex cursor-pointer p-3 justify-between border-t-2'>
                          <div>
                            <h3 className="font-semibold">{game.name}</h3>
                            <div className='text-neutral-500'>
                              {game.released.split('-')[0]} {" "}
                              {game.genres.map(genre => genre.name).join(', ')}
                            </div>
                          </div>
                          <div className='flex gap-4 items-center'>
                            <button className='rounded border-[1px] border-neutral-400 p-1 text-neutral-500
                            hover:bg-green-400 hover:text-white hover:border-green-400'>
                              $59.99</button>
                          </div>
                        </li>})}
            </ul>
          </div>}
          </>
          :
          <nav className='flex-1 animate-open hidden md:block font-semibold'>
            <ul className='flex gap-7'>
              <li className='hover:text-neutral-50 tracking-wide'> <NavLink to={'/'}> HOME </NavLink> </li>
              <li className='hover:text-neutral-50 tracking-wide'><NavLink to={'/games/rpg'}> RPG </NavLink></li>
              <li className='hover:text-neutral-50 tracking-wide'><NavLink to={'/games/strategy'}> STRATEGY </NavLink></li>
              <li className='hover:text-neutral-50 tracking-wide'><NavLink to={'/games/shooters'}> SHOOTERS </NavLink></li>
              <li className='hover:text-neutral-50 tracking-wide'><NavLink to={'/games/action'}> ACTION </NavLink></li>
            </ul>
          </nav>
          }
          <div className='flex gap-5 relative flex-1 justify-end'>
            <button onClick={toggleCart} aria-label='cart' className='z-[102] flex items-center hover:text-neutral-50'>
              <Icon path={mdiCartOutline} size={1} /> {cart.length}
            </button>
            {isSearchBarOpen ?
            <button onClick={closeSearchBar} aria-label='close-search' className='z-[101] hover:text-neutral-50'>
            <Icon path={mdiClose} size={1} />
            </button>
            :
            <button onClick={openSearchBar} aria-label='open-search' className='hover:text-neutral-50'>
              <Icon path={mdiMagnify} size={1} />
            </button>
            }
            {isMobileMenuOpen?
              <button onClick={closeMobileMenu} aria-label='close-menu' className='hover:text-neutral-50 block md:hidden'>
              <Icon path={mdiClose} size={1} />
              </button>
              :
              <button onClick={openMobileMenu} aria-label="open-menu" className='block md:hidden'>
              <Icon path={mdiMenu} size={1} />
              </button>
            }
            
              {isCartOpen &&
                <>
                <div  data-testid = "overlay" className='z-[101] fixed top-0 left-0 w-full h-full' onClick={closeCart}></div>
                  <div className='absolute bg-white p-3 top-[56px] right-[30px] shadow-md z-[103] animate-openfast
                    after:absolute after:bottom-full after:left-[89%] after:border-[5px] after:border-solid after:border-b-white after:border-t-transparent after:border-r-transparent after:border-l-transparent'>
                    {cart.length === 0 ?
                    <div className='flex flex-col items-center p-10 gap-3'>
                      <Icon path={mdiCartOutline} size={1} color={'green'}/>
                      <h3 className='text-green-600'>YOUR CART IS EMPTY</h3>
                      <hr className=' border-neutral-300 w-full'></hr>
                      <p className='text-neutral-500'>Explore great games and offers</p>
                      <button className='text-neutral-700 mt-2 border-neutral-300 border-[1px] rounded p-2 bg-neutral-200 font-semibold'>
                        BROWSE BEST SELLERS
                      </button>
                    </div>
                      :
                      <ul>
                      {cart.map((item) => {
                        return <li key={item.id} className="text-black">
                            {item.name}
                          </li>;
                      })}
                      </ul>
                    } 
                  </div>
                </>
              }
              
          </div>
        </div>
        
      </header>
      <main className='md:flex md:flex-col md:items-center overflow-x-hidden'>
        {isMobileMenuOpen ?
          <nav className='flex-1 animate-open flex justify-center 
           border-t-neutral-500 border-t-[1px] bg-neutral-900 text-neutral-300 h-screen pt-5'>
          <ul className='flex gap-7 flex-col items-center'>
            <li className='hover:text-neutral-50 tracking-wide'> <NavLink to={'/'}> HOME </NavLink> </li>
            <li className='hover:text-neutral-50 tracking-wide'><NavLink to={'/games/rpg'}> RPG </NavLink></li>
            <li className='hover:text-neutral-50 tracking-wide'><NavLink to={'/games/strategy'}> STRATEGY </NavLink></li>
            <li className='hover:text-neutral-50 tracking-wide'><NavLink to={'/games/shooters'}> SHOOTERS </NavLink></li>
            <li className='hover:text-neutral-50 tracking-wide'><NavLink to={'/games/action'}> ACTION </NavLink></li>
          </ul>
        </nav>
        :
        <Outlet />
        }
      </main>
      <footer className='bg-neutral-900 py-3 text-white text-center mt-3'>By elk15 - 2023</footer>
    </>
  )
}

Root.propTypes = {
  cart: PropTypes.array,
  allGames: PropTypes.array,
  addItemToCart: PropTypes.func
}

export default Root;
