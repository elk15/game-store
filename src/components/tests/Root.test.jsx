import {  render, screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect } from "vitest";
import { BrowserRouter } from 'react-router-dom';
import Root from "../Root";

describe("Search Bar", () => {
    afterEach(cleanup)

    it("renders search bar when search icon is clicked", async () => {
        const user = userEvent.setup();
        const fakeCart = [];
        render(
            <BrowserRouter>
                <Root cart={fakeCart}/>
            </BrowserRouter>
        );
        const btn = screen.getByLabelText('open-search');

        await user.click(btn);

        expect(screen.getByRole('searchbox')).toBeInTheDocument();
    })

    it("closes search bar when close icon is clicked", async () => {
        const user = userEvent.setup();
        const fakeCart = [];
        render(
            <BrowserRouter>
                <Root cart={fakeCart}/>
            </BrowserRouter>
        );
        const openBtn = screen.getByLabelText('open-search');

        await user.click(openBtn);

        const closeBtn = screen.getByLabelText('close-search');

        await user.click(closeBtn);

        expect(screen.getByRole('navigation')).toBeInTheDocument();
    })

    it("closes search bar when anything but it is clicked", async () => {
        const user = userEvent.setup();
        const fakeCart = [];
        render(
            <BrowserRouter>
                <Root cart={fakeCart}/>
            </BrowserRouter>
        );
        const openBtn = screen.getByLabelText('open-search');

        await user.click(openBtn);

        const closeBtn = screen.getByTestId('overlay');

        await user.click(closeBtn);

        expect(screen.getByRole('navigation')).toBeInTheDocument();
    })

    it("search displays results correctly", async () => {
        const user = userEvent.setup();
        const allGames = [{id : 1, name: 'The Game', released: '2022-01-02', genres: [{name: 'RPG'}]}];
        const fakeCart = [];
        render(
            <BrowserRouter>
                <Root cart={fakeCart} allGames={allGames}/>
            </BrowserRouter>
        );
        const openBtn = screen.getByLabelText('open-search');

        await user.click(openBtn);
        const searchInput = screen.getByRole('searchbox')
        await user.type(searchInput, 'The');

        expect(screen.getByText("The Game")).toBeInTheDocument();
    })
})

describe("Cart Modal", () => {
    afterEach(cleanup)


    it("renders cart modal when cart is clicked", async () => {
        const user = userEvent.setup();
        const fakeCart = [];
        render(
            <BrowserRouter>
                <Root cart={fakeCart}/>
            </BrowserRouter>
        );
        const btn = screen.getByLabelText('cart');

        await user.click(btn);

        expect(screen.getByTestId('overlay')).toBeInTheDocument();
    })

    it("closes cart modal when cart is clicked", async () => {
        const user = userEvent.setup();
        const fakeCart = [];
        render(
            <BrowserRouter>
                <Root cart={fakeCart}/>
            </BrowserRouter>
        );
        const openBtn = screen.getByLabelText('cart');

        await user.click(openBtn);

        const closeBtn = screen.getByLabelText('cart');

        await user.click(closeBtn);

        expect(screen.queryAllByTestId('overlay').length).toBe(0);
    })

    it("closes cart modal when anything but it is clicked", async () => {
        const user = userEvent.setup();
        const fakeCart = [];
        render(
            <BrowserRouter>
                <Root cart={fakeCart}/>
            </BrowserRouter>
        );
        const openBtn = screen.getByLabelText('cart');

        await user.click(openBtn);

        const closeBtn = screen.getByTestId('overlay');

        await user.click(closeBtn);

        expect(screen.queryAllByTestId('overlay').length).toBe(0);
    })

    it("displays cart items", async () => {
        const user = userEvent.setup();
        const fakeCart = [{id: 1, name: 'My Game', background_image: ''}];
        render(
            <BrowserRouter>
                <Root cart={fakeCart}/>
            </BrowserRouter>
        );
        const openBtn = screen.getByLabelText('cart');

        await user.click(openBtn);

        expect(screen.getByText('My Game')).toBeInTheDocument();
    })

    it("displays appropriate text if cart is empty", async () => {
        const user = userEvent.setup();
        const fakeCart = [];
        render(
            <BrowserRouter>
                <Root cart={fakeCart}/>
            </BrowserRouter>
        );
        const openBtn = screen.getByLabelText('cart');

        await user.click(openBtn);

        expect(screen.getByText('YOUR CART IS EMPTY')).toBeInTheDocument();
    })
})