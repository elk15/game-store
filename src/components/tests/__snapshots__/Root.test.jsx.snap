// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > renders Header 1`] = `
<DocumentFragment>
  <header
    class="z-[100] opacity-95 flex justify-center bg-neutral-900  py-4 px-3 text-neutral-300 sticky top-0 text-sm "
  >
    <div
      class="max-w-screen-lg flex flex-1 gap-3 items-center justify-between"
    >
      <nav
        class="flex-1 animate-open hidden md:block font-semibold"
      >
        <ul
          class="flex gap-7"
        >
          <li
            class="hover:text-neutral-50 tracking-wide"
          >
             
            <a
              aria-current="page"
              class="active"
              href="/"
            >
               HOME 
            </a>
             
          </li>
          <li
            class="hover:text-neutral-50 tracking-wide"
          >
            <a
              class=""
              href="/games/rpg"
            >
               RPG 
            </a>
          </li>
          <li
            class="hover:text-neutral-50 tracking-wide"
          >
            <a
              class=""
              href="/games/strategy"
            >
               STRATEGY 
            </a>
          </li>
          <li
            class="hover:text-neutral-50 tracking-wide"
          >
            <a
              class=""
              href="/games/shooters"
            >
               SHOOTERS 
            </a>
          </li>
          <li
            class="hover:text-neutral-50 tracking-wide"
          >
            <a
              class=""
              href="/games/action"
            >
               ACTION 
            </a>
          </li>
        </ul>
      </nav>
      <div
        class="flex gap-5 relative flex-1 justify-end"
      >
        <button
          aria-label="cart"
          class="flex items-center hover:text-neutral-50"
        >
          <svg
            role="presentation"
            style="width: 1.5rem; height: 1.5rem;"
            viewBox="0 0 24 24"
          >
            <path
              d="M17,18A2,2 0 0,1 19,20A2,2 0 0,1 17,22C15.89,22 15,21.1 15,20C15,18.89 15.89,18 17,18M1,2H4.27L5.21,4H20A1,1 0 0,1 21,5C21,5.17 20.95,5.34 20.88,5.5L17.3,11.97C16.96,12.58 16.3,13 15.55,13H8.1L7.2,14.63L7.17,14.75A0.25,0.25 0 0,0 7.42,15H19V17H7C5.89,17 5,16.1 5,15C5,14.65 5.09,14.32 5.24,14.04L6.6,11.59L3,4H1V2M7,18A2,2 0 0,1 9,20A2,2 0 0,1 7,22C5.89,22 5,21.1 5,20C5,18.89 5.89,18 7,18M16,11L18.78,6H6.14L8.5,11H16Z"
              style="fill: currentColor;"
            />
          </svg>
           0
        </button>
        <button
          aria-label="open-search"
          class="hover:text-neutral-50"
        >
          <svg
            role="presentation"
            style="width: 1.5rem; height: 1.5rem;"
            viewBox="0 0 24 24"
          >
            <path
              d="M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z"
              style="fill: currentColor;"
            />
          </svg>
        </button>
        <button
          aria-label="open-menu"
          class="block md:hidden"
        >
          <svg
            role="presentation"
            style="width: 1.5rem; height: 1.5rem;"
            viewBox="0 0 24 24"
          >
            <path
              d="M3,6H21V8H3V6M3,11H21V13H3V11M3,16H21V18H3V16Z"
              style="fill: currentColor;"
            />
          </svg>
        </button>
      </div>
    </div>
  </header>
  <main
    class="md:flex md:flex-col md:items-center overflow-x-hidden"
  />
  <footer
    class="bg-neutral-900 py-3 text-white text-center mt-3"
  >
    By elk15 - 2023
  </footer>
</DocumentFragment>
`;
